### PR TITLE
[varLib] add support for building variable COLR from set of master COLRv1 tables

### PR DESCRIPTION
**Pull request by anthrotype**
Fixes https://github.com/fonttools/fonttools/issues/2328

I'd like to add more unit tests before merge, but since this is already been used for generating [variable COLRv1 test fonts](https://github.com/googlefonts/color-fonts/pull/102) to aid Skia/FreeType implementation, as well as for stuff like https://github.com/BlackFoundryCom/black-renderer/pull/88, I think it's better to put this out for review so I stop tinkering with it.

### CODE REVIEW

File: Lib/fontTools/varLib/__init__.py [hunk 1]
```diff
@@ -711,6 +713,28 @@ def _add_CFF2(varFont, model, master_fonts):
...
+	merger = COLRVariationMerger(model, axisTags, font)
+	merger.mergeTables(font, master_fonts)
+	store = merger.store_builder.finish()
+
+	colr = font["COLR"].table
+	if store.VarData:
+		mapping = store.optimize()
+		colr.VarStore = store
+		# the special 0xFFFF/0xFFFF (for no variations) always maps to itself
+		mapping[ot.NO_VARIATION_INDEX] = ot.NO_VARIATION_INDEX
```
Comment (behdad) [inline Lib/fontTools/varLib/__init__.py:726]: This needs to go into the VarStore logic itself.
Generally this issue: https://github.com/fonttools/fonttools/issues/2211

Comment (behdad) [reply Lib/fontTools/varLib/__init__.py:726]: Ie. Other places are affected as well.

**Comment by behdad** [comment]
I reviewed this. Great work @anthrotype. Some of the changes to otBase machinery are a bit crude but I don't see any cleaner way. Thanks.

**Comment by anthrotype** [comment]
reminder to self: given how with variability, pre-computing tight clip boxes is non-trivial (think of rotate, skew or other complex transforms), we may have to add an option to drop the clip boxes in the variable COLR (they are optional, renderer can compute on the fly from the paint graph), or better use some euristics to decide if it's safe or not to keep them on a glyph-by-glyph basis. Right now they are simply linearly interpolated like everything else.

Also see https://github.com/googlefonts/colr-gradients-spec/issues/289

**Comment by anthrotype** [comment]
one notable change since my last push. Previously, the merger required that all masters had the same number of BaseGlyphList records, as well as the same LayerList count, and would fail as soon as it found a PaintColrLayers with a different FirstLayerIndex. But that immediately ruled out masters with "sparse" color glyph sets, which was too restrictive, given how varLib.build does support that when merging, say, gvar variations, for example.
Now, with the latest changes (esp. https://github.com/fonttools/fonttools/pull/2660/commits/e5029801d2c1739b8a51de926b3cb636edf2926e), it is possible to merge multiple master COLR tables containing different base color glyphs, or having different LayerList count or layer-reuse layout. Instead of treating PaintColrLayers as constant tables, I now traverse their references into the LayerList while merging the paint graph rooted at each base color glyph, and only require that each color glyph masters resolve to the same number of (flattened) layers. For the same reason, it's no longer necessary (even though it may speed things up a bit) to ensure that layer reuse is disabled when building each master COLR table, the algorithm works either way (duplicating references as needed to make sure the merger's in-place mutation doesn't go bad). And it's no longer necessary to unbuild the variable COLR table at the end just to be able to rebuild it with layer reuse turned on, since now the merger applies layer reuse as it goes along.

**Pull request merged**